packages/apollo-engine-reporting/src/schemaReporter.ts

```diff
@@ -0,0 +1,164 @@
+// Fields required for the trial related emails
+import {
+  ReportServerInfoVariables,
+  EdgeServerInfo,
+  AutoregReportServerInfoResult,
+} from './reportingOperationTypes';
+import { fetch, Headers, Request } from 'apollo-server-env';
+import { GraphQLRequest, Logger } from 'apollo-server-types';
+
+const reportServerInfoGql = `
+  mutation ReportServerInfo($info: EdgeServerInfo!, $executableSchema: String) {
+    me {
+      __typename
+      ... on ServiceMutation {
+        reportServerInfo(info: $info, executableSchema: $executableSchema) {
+          inSeconds
+          withExecutableSchema
+        }
+      }
+    }
+  }
+`;
+
+class ReportingError extends Error {}
+
+export function reportingLoop(
+  schemaReporter: SchemaReporter,
+  logger: Logger,
+  sendNextWithExecutableSchema: boolean,
+  fallbackReportingDelayInMs: number,
+) {
+  function inner() {
+    // Bail out permanently
+    if (schemaReporter.stopped()) return;
+
+    schemaReporter
+      .reportServerInfo(sendNextWithExecutableSchema)
+      .then(({ inSeconds, withExecutableSchema }) => {
+        sendNextWithExecutableSchema = withExecutableSchema;
+        setTimeout(inner, inSeconds * 1000);
+      })
+      .catch((error: any) => {
+        // In the case of an error we want to continue looping
+        // We can add hardcoded backoff in the future,
+        // or on repeated failures stop responding reporting.
+        logger.warn(`Error in reportingServerInfo: ${error}`);
+        sendNextWithExecutableSchema = false;
+        setTimeout(inner, fallbackReportingDelayInMs);
+      });
+  }
+
+  inner();
+}
+
+interface ReportServerInfoReturnVal {
+  inSeconds: number;
+  withExecutableSchema: boolean;
+}
+
+// This class is meant to be a thin shim around the gql mutations.
+export class SchemaReporter {
+  // These mirror the gql variables
+  private readonly serverInfo: EdgeServerInfo;
+  private readonly executableSchemaDocument: any;
+  private readonly url: string;
+
+  private isStopped: boolean;
+  private readonly headers: Headers;
+
+  constructor(
+    serverInfo: EdgeServerInfo,
+    schemaSdl: string,
+    apiKey: string,
+    reportingEndpoint: string | undefined,
+  ) {
+    if (apiKey === '') {
+      throw new Error('No api key defined');
+    }
+
+    this.headers = new Headers();
+    this.headers.set('Content-Type', 'application/json');
```

Comment:
Let's please add an `apollographql-client-name` and `apollographql-client-version` header